### ISSUE TITLE
PTA Problem with function pointers within C structure

### ISSUE THREAD

**Issue by h1994st** [mchalupa/dg#312]
Hi,

I tried to slice a program against the criteria of `a_func` (See [example.c](https://gist.github.com/h1994st/6441548d0fec7a48f84dfe27059f973f#file-example-c)), which is called by a function pointer stored in a C structure. However, in the sliced program, all instructions in `main` function are removed (see [example-debug.ll](https://gist.github.com/h1994st/6441548d0fec7a48f84dfe27059f973f#file-example-debug-ll)).

I use `llvm-pta-dump` to insepct the pointer graph. It seems PTA cannot properly handle the function pointer in the structure (see Line 24&25 of [example.pta-dump.log](https://gist.github.com/h1994st/6441548d0fec7a48f84dfe27059f973f#file-example-pta-dump-log-L24-L25)).

Any hints to solve this problem?

Thank you!

**Comment by mchalupa** [comment]
Thanks for the report. It looks indeed as a bug.

**Comment by mchalupa** [comment]
It looks that the global structure `test_instance_2` is not properly initialized.

**Comment by mchalupa** [comment]
Or rather a bug in getelementptr handling -- the `test_instance_2` structure has `a_func` stored on offset 12 (which should be correct), but the getelementptr that is used to get the address for loading the function pointer has the offset 16.

**Comment by h1994st** [comment]
> Or rather a bug in getelementptr handling -- the `test_instance_2` structure has `a_func` stored on offset 12 (which should be correct), but the getelementptr that is used to get the address for loading the function pointer has the offset 16.

Hi @mchalupa,

I think the correct offset is 16 due to the aliment within the structure. The `getelementptr` handling is correct. The fundamental reason for the reported issue is the incorrect offset in `handleGlobalVariableInitializer`. I just submitted a pull request (#313) to fix this issue.

Thank you!

---

Here is a detailed description:

In `createGEP` (shown below), `accumulateConstantOffset` is invoked to get the offset. I took a look at the source code of that function ([link](https://github.com/llvm/llvm-project/blob/89de0d8dfbb9a6ff1f8b141ed70b563ecc094878/llvm/lib/IR/Operator.cpp#L35-L63)). For structure type, LLVM uses the function `getElementOffset` of `StructureLayout` to determine the offset rather than function `getTypeAllocSize` of `DataLayout`, which reminds me of the alignment with the structure.

https://github.com/mchalupa/dg/blob/dce8e81c07fea830fcadce4ee8c9548434c658e9/lib/llvm/analysis/PointsTo/Instructions.cpp#L64-L65

I then checked the actual size of the structure by printing the size and the result is 24.

```c
printf("%lu\n", sizeof(struct test_type_t));
```

Thus, I modified the function `handleGlobalVariableInitializer` in `Globals.cpp` to fix the offset.
